// @fsoc/royalroadl-api/dist/services/chapter.js
import logger from 'debug';
import * as cheerio from 'cheerio';
import { getLastPage } from '../utils.js';
import { RoyalError, RoyalResponse } from '../responses.js';
import date from 'date.js';
export class ChapterService {
    static isValidNewChapter(chapter) {
        return chapter.content.length >= 500 && chapter.title.length <= 150;
    }
    debug;
    req;
    constructor(req) {
        this.req = req;
        this.debug = logger('chapter');
    }
    async publish(fictionID, chapter) {
        this.requireAuth();
        if (!ChapterService.isValidNewChapter(chapter)) {
            throw new RoyalError('Invalid chapter');
        }
        await this.req.post(`/fiction/chapter/new/${String(fictionID)}`, {
            Status: 'New',
            fid: fictionID.toString(),
            Title: chapter.title,
            Content: chapter.content,
            PreAuthorNotes: chapter.preNote || '',
            PostAuthorNotes: chapter.postNote || '',
            action: 'publish',
        }, { fetchToken: true });
        return new RoyalResponse(chapter);
    }
    async getChapter(chapterID) {
        const body = await this.req.get(`/fiction/0/_/chapter/${String(chapterID)}/_`);
        const chapter = ChapterParser.parseChapter(body);
        return new RoyalResponse(chapter);
    }
    async getComments(chapterID, page = 1) {
        const path = `/fiction/chapter/${String(chapterID)}/comments/`;
        const lastPage = page === 'last' ? getLastPage(await this.req.get(path)) : page;
        const body = await this.req.get(path + lastPage);
        const comments = ChapterParser.parseComments(body);
        return new RoyalResponse(comments);
    }
    async postComment(chapterID, content) {
        this.requireAuth();
        await this.req.post(`/fiction/0/_/chapter/${String(chapterID)}/_`, {
            cid: chapterID.toString(),
            action: 'submit',
            comment: content
                .split('\n')
                .map((e) => `<p>${e}</p>`)
                .join(''),
        }, { fetchToken: true, ignoreStatus: true });
        return new RoyalResponse(null);
    }
    requireAuth() {
        if (!this.req.isAuthenticated) {
            throw new RoyalError('Not authenticated.');
        }
    }
}
class ChapterParser {
    static parseChapter(html) {
        const $ = cheerio.load(html);
        const notes = $('div.author-note');
        const preNote = $(notes).eq(0).find('p').text();
        const postNote = $(notes).eq(1).find('p').text();
        const content = ($('div.chapter-inner.chapter-content').html() || '').trim();
        const next = ChapterParser.getChapterID($('i.fa-chevron-double-right').parent().attr('href') || '');
        const previous = ChapterParser.getChapterID($('i.fa-chevron-double-left').parent().attr('href') || '');
        return { content, preNote, postNote, next, previous };
    }
    static parseComments(html) {
        const $ = cheerio.load(html);
        const comments = [];
        $('div.media.media-v2').each((i, el) => {
            const id = parseInt($(el).attr('id')?.split('-')[2] ?? '', 10);
            const posted = date($(el).find('time').text() + ' ago').getTime();
            let raw = '';
            $(el)
                .find('div.media-body')
                .find('p')
                .each((j, p) => {
                const text = $(p).text();
                if (text) {
                    raw += text.trim() + '\n';
                }
            });
            const content = raw.trim();
            const author = {
                avatar: $(el).find('img').attr('src') ?? '',
                name: $(el).find('span.name').find('a').text().split('@')[0],
                id: parseInt($(el).find('span.name').find('a').attr('href')?.split('/')[2] ?? '', 10),
            };
            comments.push({ id, author, posted, content });
        });
        return comments;
    }
    static getChapterID(url) {
        const split = url.split('/');
        return split.length < 5 ? -1 : parseInt(split[5], 10);
    }
}
