// @fsoc/royalroadl-api/dist/services/fictions.js
import * as cheerio from 'cheerio';
import { RoyalError, RoyalResponse } from '../responses.js';
import date from 'date.js';
const statsKeys = ['pages', 'latest', 'rating', 'chapters', 'followers'];
export class FictionsService {
    req;
    constructor(req) {
        this.req = req;
    }
    async getLatest(page = 1) {
        const body = await this.getList('latest-updates', page);
        const fictions = FictionsParser.parseLatest(body);
        return new RoyalResponse(fictions);
    }
    async getPopular(page = 1) {
        const body = await this.getList('active-popular', page);
        const fictions = FictionsParser.parsePopular(body);
        return new RoyalResponse(fictions);
    }
    async getBest(page = 1) {
        const body = await this.getList('best-rated', page);
        const fictions = FictionsParser.parsePopular(body);
        return new RoyalResponse(fictions);
    }
    async getNewReleases(page = 1) {
        const body = await this.getList('new-releases', page);
        const fictions = FictionsParser.parseNewReleases(body);
        return new RoyalResponse(fictions);
    }
    async search(query, page = 1) {
        const body = await this.getList(`search?${query}`, page);
        const fictions = FictionsParser.parseSearch(body);
        return new RoyalResponse(fictions);
    }
    async getList(type, page = 1) {
        const path = `/fictions/${type}`;
        const body = await this.req.get(path, { page: String(page) });
        function validPage(html) {
            return !html.includes('There is nothing here :(');
        }
        if (!validPage(body)) {
            throw new RoyalError('No fictions found.');
        }
        else {
            return body;
        }
    }
}
class FictionsParser {
    static parseLatest(html) {
        const $ = cheerio.load(html);
        const fictions = [];
        $('.fiction-list-item').each((i, el) => {
            const latest = [];
            $(el)
                .find('li.list-item')
                .each((j, item) => {
                latest.push({
                    link: $(item).find('a').attr('href') ?? '',
                    name: $(item).find('span.col-xs-8').text(),
                    created: date($(item).find('time').text()).getTime(),
                });
            });
            fictions.push(Object.assign(FictionsParser.parseBlurb($, el), { latest }));
        });
        return fictions;
    }
    static parsePopular(html) {
        const $ = cheerio.load(html);
        const fictions = [];
        $('.fiction-list-item').each((i, el) => {
            let description = '';
            $(el)
                .find('.margin-top-10.col-xs-12')
                .find('p')
                .each((j, para) => {
                description += $(para).text() + '\n';
            });
            const stats = {
                pages: 0,
                latest: 0,
                rating: 0,
                chapters: 0,
                followers: 0,
            };
            stats.latest = date($(el).find('time').attr('datetime') ?? '').getTime();
            stats.rating = parseFloat($(el).find('.star').attr('title') ?? '');
            $(el)
                .find('span')
                .each((j, stat) => {
                const text = $(stat).text().toLowerCase();
                const key = text.split(' ')[1];
                const value = parseInt(text.split(' ')[0].replace(/,/gi, ''), 10);
                if (!key || !value) {
                    return;
                }
                if (statsKeys.includes(key)) {
                    stats[key] = value;
                }
            });
            fictions.push(Object.assign(FictionsParser.parseBlurb($, el), { description }, { stats }));
        });
        return fictions;
    }
    static parseSearch(html) {
        const $ = cheerio.load(html);
        const fictions = [];
        $('.fiction-list-item').each((i, el) => {
            const image = $(el).find('img').attr('src') ?? '';
            const titleEl = $(el).find('h2.fiction-title').children('a');
            const title = $(titleEl).text();
            const id = parseInt($(titleEl).attr('href')?.split('/')[2] ?? '', 10);
            const pages = parseInt($(el).find('i.fa-book').next().text().replace('Pages', '').trim(), 10);
            let description = '';
            $(el)
                .find('div.margin-top-10.col-xs-12')
                .children()
                .each((i, el) => {
                description += $(el).text();
            });
            fictions.push({ id, title, pages, image, description });
        });
        return fictions;
    }
    static parseNewReleases(html) {
        const $ = cheerio.load(html);
        const fictions = [];
        $('div.fiction-list-item').each((i, el) => {
            let description = '';
            $(el)
                .find('div.hidden-content')
                .find('p')
                .each((j, p) => {
                description += $(p).text().trim() + '\n';
            });
            description = description.trim();
            fictions.push(Object.assign(FictionsParser.parseBlurb($, el), { description }));
        });
        return fictions;
    }
    static parseBlurb($, el) {
        const titleEl = $(el).find('.fiction-title').children('a');
        const title = $(titleEl).text();
        const image = $(el).find('img').attr('src') ?? '';
        const type = $(el).find('span.label.bg-blue-hoki').text();
        const id = parseInt($(titleEl).attr('href')?.split('/')[2] ?? '', 10);
        const tags = $(el)
            .find('a.label.fiction-tag')
            .map((i, tag) => $(tag).text())
            .get();
        return { id, type, tags, title, image };
    }
}
