// @fsoc/royalroadl-api/dist/services/fiction.js
import * as cheerio from 'cheerio';
import { getLastPage } from '../utils.js';
import { RoyalResponse } from '../responses.js';
import date from 'date.js';
export class FictionService {
    req;
    constructor(req) {
        this.req = req;
    }
    async getFiction(id) {
        const path = `/fiction/${String(id)}`;
        const body = await this.req.get(path);
        const fiction = FictionParser.parseFiction(body);
        return new RoyalResponse(fiction);
    }
    async getRandom() {
        const path = `/fiction/random`;
        const body = await this.req.get(path);
        const fiction = FictionParser.parseFiction(body);
        return new RoyalResponse(fiction);
    }
    async getReviews(id, page = 1) {
        const path = `/fiction/${String(id)}`;
        const body = await this.req.get(path, {
            reviews: String(page === 'last' ? getLastPage(await this.req.get(path)) : page),
        });
        const reviews = FictionParser.parseReviews(body);
        return new RoyalResponse(reviews);
    }
}
class FictionParser {
    static parseFiction(html) {
        const $ = cheerio.load(html);
        const title = $('div.fic-title').find('h1').text();
        const image = $('div.fic-header').find('img').attr('src') ?? '';
        const labels = $('span.bg-blue-hoki');
        const type = labels.eq(0).text();
        const status = labels.eq(1).text().trim();
        const tags = $('span.tags')
            .find('a.label')
            .map((i, el) => $(el).text().trim())
            .get();
        const warnings = $('ul.list-inline')
            .find('li')
            .map((i, el) => $(el).text().trim())
            .get();
        const description = $('div.description > div.hidden-content').text().trim();
        const authorEl = $('.portlet-body').eq(0);
        const author = {
            name: $(authorEl).find('.mt-card-content').find('a').text().trim(),
            title: $(authorEl).find('.mt-card-desc').text(),
            avatar: $(authorEl).find('img[data-type="avatar"]').attr('src') ?? '',
            id: +($(authorEl).find('.mt-card-content').find('a').attr('href') || '').split('/')[2],
        };
        const statsEl = $('div.stats-content');
        const statsList = $(statsEl).find('.list-unstyled').eq(1).find('li');
        const ratingList = $(statsEl).find('.list-unstyled').eq(0).find('li');
        const parseNumber = (raw) => parseInt(raw.replace(/,/gi, ''), 10);
        const parseRating = (raw) => typeof raw === 'string' ? parseFloat(raw.split('/')[0].trim()) : -1;
        const getContent = (el) => el.find('span').data('content');
        const stats = {
            pages: parseNumber($(statsList).eq(11).text()),
            ratings: parseNumber($(statsList).eq(9).text()),
            followers: parseNumber($(statsList).eq(5).text()),
            favorites: parseNumber($(statsList).eq(7).text()),
            views: {
                total: parseNumber($(statsList).eq(1).text()),
                average: parseNumber($(statsList).eq(3).text()),
            },
            score: {
                style: parseRating(getContent($(ratingList).eq(3))),
                story: parseRating(getContent($(ratingList).eq(5))),
                grammar: parseRating(getContent($(ratingList).eq(9))),
                overall: parseRating(getContent($(ratingList).eq(1))),
                character: parseRating(getContent($(ratingList).eq(7))),
            },
        };
        const chapters = [];
        $('tbody')
            .find('tr')
            .each((i, el) => {
            chapters.push({
                title: $(el).find('td').eq(0).find('a').text().trim(),
                id: parseInt($(el).find('td').eq(0).find('a').attr('href')?.split('/')[5] ?? '', 10),
                release: date($(el).find('td').eq(1).find('time').text()).getTime(),
            });
        });
        return {
            type,
            tags,
            stats,
            title,
            image,
            status,
            author,
            warnings,
            chapters,
            description,
        };
    }
    static parseReviews(html) {
        const $ = cheerio.load(html);
        const reviews = [];
        $('div.review').each((i, el) => {
            const posted = parseInt($(el).find('time').attr('unixtime') ?? '', 10);
            let content = '';
            $(el)
                .find('div.review-content')
                .each((j, p) => {
                content += $(p).text() + '\n';
            });
            content = content.trim();
            const authorLink = $(el).find('div.review-meta').find('a').eq(0);
            const author = {
                name: $(authorLink).text(),
                avatar: $(el).find('img').attr('src') ?? '',
                id: parseInt($(authorLink).attr('href')?.split('/')[2] ?? '', 10),
            };
            const overallScore = parseFloat($(el)
                .find('div.overall-score-container')
                .children()
                .filter((i, el) => $(el).attr('aria-label')?.includes('stars') ?? false)
                .first()
                .attr('aria-label')
                ?.replace('stars', '')
                .trim() ?? '') || 0;
            const score = {
                style: 0,
                story: 0,
                overall: overallScore,
                grammar: 0,
                character: 0,
            };
            reviews.push({ score, author, content, posted });
        });
        return reviews;
    }
}
