// @fsoc/royalroadl-api/dist/services/user.js
import * as cheerio from 'cheerio';
import { getBaseAddress } from '../constants.js';
import { RoyalError, RoyalResponse } from '../responses.js';
import { isAd } from '../utils.js';
import date from 'date.js';
export class UserService {
    req;
    constructor(req) {
        this.req = req;
    }
    get isLoggedIn() {
        return this.req.isAuthenticated;
    }
    async login(email, password) {
        await this.req.post('/account/login', { Email: email, Password: password, ReturnUrl: '/home', Remember: 'true' }, {
            fetchToken: true,
            ignoreStatus: true,
        });
        return new RoyalResponse('Logged in.');
    }
    async getMyFictions() {
        if (!this.isLoggedIn) {
            throw new RoyalError('Not authenticated.');
        }
        const body = await this.req.get('/my/fictions');
        const myFictions = UserParser.parseMyFictions(body);
        return new RoyalResponse(myFictions);
    }
    async getMyFollows(page = 1) {
        if (!this.isLoggedIn) {
            throw new RoyalError('Not authenticated.');
        }
        const body = await this.req.get('/my/follows', { page: String(page) });
        const $ = cheerio.load(body);
        const hasContent = $('div.fiction-list-item').length > 0;
        if (!hasContent) {
            throw new RoyalError('Out of bounds.');
        }
        const myBookmarks = UserParser.parseMyFollows(body);
        return new RoyalResponse(myBookmarks);
    }
    async getNotifications() {
        if (!this.isLoggedIn) {
            throw new RoyalError('Not authenticated.');
        }
        const body = await this.req.get('/notifications/get');
        const notifications = UserParser.parseNotifications(body);
        return new RoyalResponse(notifications);
    }
    async getMyReadLater(page = 1) {
        if (!this.isLoggedIn) {
            throw new RoyalError('Not authenticated.');
        }
        const body = await this.req.get('/my/readlater', { page: String(page) });
        const $ = cheerio.load(body);
        const hasContent = $('div.fiction-list-item').length > 0;
        if (!hasContent) {
            throw new RoyalError('Out of bounds.');
        }
        const myBookmarks = UserParser.parseMyReadLater(body);
        return new RoyalResponse(myBookmarks);
    }
}
class UserParser {
    static parseMyFollows(html) {
        const $ = cheerio.load(html);
        const myBookmarks = [];
        $('div.fiction-list-item').each((i, el) => {
            if (isAd($(el))) {
                return;
            }
            const image = $(el).find('img').attr('src');
            if (!image) {
                return;
            }
            const titleEl = $(el).find('h2.fiction-title').find('a');
            const title = $(titleEl).text().trim();
            const titleHref = $(titleEl).attr('href');
            if (!titleHref) {
                return;
            }
            const id = parseInt(titleHref.split('/')[2], 10);
            const link = $(el).find('a').attr('href');
            if (!link) {
                return;
            }
            const authorEl = $(el).find('span.author').find('a');
            const authorHref = $(authorEl).attr('href');
            if (!authorHref) {
                return;
            }
            const author = {
                name: $(authorEl).text(),
                id: parseInt(authorHref.split('/')[2], 10),
            };
            myBookmarks.push({ id, title, image, link, author });
        });
        return myBookmarks;
    }
    static parseMyFictions(html) {
        const $ = cheerio.load(html);
        const fictions = [];
        $('div.fiction').each((i, el) => {
            const title = $(el).find('h4.col-sm-10').text();
            const href = $(el).find('a').attr('href');
            if (!href) {
                return;
            }
            const id = parseInt(href.split('/')[2], 10);
            fictions.push({ id, title });
        });
        return fictions;
    }
    static parseNotifications(html) {
        const $ = cheerio.load(html);
        const notifications = [];
        $('li').each((i, el) => {
            const image = $(el).find('img').attr('src');
            if (!image) {
                return;
            }
            const message = $(el).find('span').eq(0).text().trim();
            const link = getBaseAddress() + $(el).find('a').attr('href');
            const timestamp = date($(el).find('time').text() + ' ago').getTime();
            notifications.push({ link, image, message, timestamp });
        });
        return notifications;
    }
    static parseMyReadLater(html) {
        const $ = cheerio.load(html);
        const myReadLater = [];
        $('div.fiction-list-item').each((i, el) => {
            if (isAd($(el))) {
                return;
            }
            const image = $(el).find('img').attr('src');
            if (!image) {
                return;
            }
            const titleEl = $(el).find('h2.fiction-title').find('a');
            const title = $(titleEl).text().trim();
            const titleHref = $(titleEl).attr('href');
            if (!titleHref) {
                return;
            }
            const id = parseInt(titleHref.split('/')[2], 10);
            const link = $(el).find('a').attr('href');
            if (!link) {
                return;
            }
            const pages = parseInt($(el).find('span.page-count').text().replace('pages', '').trim(), 10);
            const authorEl = $(el).find('span.author').find('a');
            const authorHref = $(authorEl).attr('href');
            if (!authorHref) {
                return;
            }
            const author = {
                name: $(authorEl).text(),
                id: parseInt(authorHref.split('/')[2], 10),
            };
            myReadLater.push({ id, title, image, link, pages, author });
        });
        return myReadLater;
    }
}
