// @fsoc/royalroadl-api/dist/requester.js
import { getBaseAddress, getUserAgent } from './constants.js';
import * as cheerio from 'cheerio';
import { RoyalError } from './responses.js';
import debug from 'debug';
import { CookieJar } from 'tough-cookie';
import got from 'got';
export class Requester {
    static headers = {
        'User-Agent': getUserAgent(),
    };
    static catchGenericErrorLegacy(html) {
        const $ = cheerio.load(html);
        const error = $('div.page-404').find('h3').text().trim();
        const alert = $('div.alert.alert-danger').eq(0).text().trim();
        return error || alert || null;
    }
    url = getBaseAddress();
    debug = debug('royalroad-api:requester');
    cookieJar = new CookieJar();
    get isAuthenticated() {
        const authenticationCookie = this.cookieJar
            .getCookiesSync(this.url)
            .find((cookie) => cookie.key === '.AspNetCore.Identity.Application');
        return authenticationCookie != null;
    }
    async get(path, data = {}, options = {}) {
        const query = new URLSearchParams(data);
        return await this.request({
            url: this.url + path + (query ? '?' + query : ''),
        }, options);
    }
    async post(path, data, options = {}) {
        this.debug('sending POST to %o with %o options', path, options);
        if (options.fetchToken) {
            data.__RequestVerificationToken = await this.fetchToken(path);
        }
        return await this.request({
            url: this.url + path,
            form: data,
            method: 'POST',
        }, options);
    }
    async request(request, internalOptions = {}) {
        this.debug('%o > %o', request.method || 'GET', request.url);
        request.cookieJar = this.cookieJar;
        request.headers = Requester.headers;
        try {
            const response = await got(request);
            this.debug('%o < %o (%o)', request.method || 'GET', response.statusCode, response.statusMessage);
            if (response.statusCode !== (internalOptions.successStatus || 200) &&
                !internalOptions.ignoreStatus) {
                throw new RoyalError(`Request error: ${response.statusMessage}`);
            }
            const genericError = this.catchGenericError(response.body);
            if (!internalOptions.ignoreParser && genericError !== null) {
                throw new RoyalError(genericError);
            }
            return response.body;
        }
        catch (error) {
            this.debug('request error %O', error);
            throw new RoyalError(error instanceof Error ? error.message : 'Unkown error');
        }
    }
    async fetchToken(path) {
        const body = await this.get(path);
        const $ = cheerio.load(body);
        const genericError = this.catchGenericError(body);
        if (genericError !== null) {
            throw new RoyalError(genericError);
        }
        const token = $('input[name="__RequestVerificationToken"]').prop('value');
        this.debug('got token %o', token);
        if (token && token.length !== 0) {
            return token;
        }
        else {
            throw new RoyalError('Token not found.');
        }
    }
    catchGenericError(html) {
        const legacy = Requester.catchGenericErrorLegacy(html);
        const $ = cheerio.load(html);
        const error = $('div.text-danger').find('li').first().text().trim() || null;
        this.debug('parsed page for generic errors %O', { error, legacy });
        return error || legacy;
    }
}
